src/components/data/DataTable.tsx

```diff
@@ -15,15 +16,20 @@ import TableFoot from './TableFoot';
 import TableHead from './TableHead';
 import TableRow from './TableRow';
 
-export type TableColumn = {
-  field: string;
+export type TableColumn<Field> = {
+  field: Field;
   label: string;
   classes?: string;
 };
 
+export type Order<Field> = {
+  field: Field;
+  direction: 'ascending' | 'descending';
+} | null;
+
 type ComponentProps<Row> = {
   rows: Row[];
-  columns: TableColumn[];
+  columns: TableColumn<keyof Row>[];
```

Comment:
Making this generic requires being more strict in downstream projects, but it's not critical. I tested this in LMS, and the only requirement was typing the array of columns as `DataTableProps<foo>['columns']`.

In the docs I used a different approach, which is setting the values `as const` on the fly. Both options should work and we might choose one over the other depending on the case.

Other than that, it is very useful to avoid typos, and plays well with the new `orderableColumns` prop.

Reply:
This assumes that the `Row` objects are always going to be objects with properties that match the columns. I expect that is usually true. It might turn out to be too strict a requirement in some cases. An alternative would be to define a second generic for `ComponentProps` which is the type of columns, defaulting to `keyof Row`.

Reply:
> An alternative would be to define a second generic for ComponentProps which is the type of columns, defaulting to keyof Row.

I like this idea.

Reply:
I tried to apply a less strict logic, but TypeScript ends up complaining one way or another because the logic assumes a correlation between columns and the fields on every row for different tasks. In fact, I realized we are already using `keyof Row` at least in two places. See [line 63](https://github.com/hypothesis/frontend-shared/blob/8c9a30c6dbf03e883a33b94212a199d65744d22f/src/components/data/DataTable.tsx#L63) and [line 74](https://github.com/hypothesis/frontend-shared/blob/8c9a30c6dbf03e883a33b94212a199d65744d22f/src/components/data/DataTable.tsx#L74).

On top of that, when it comes to logic related specifically with ordering, a new generic type that defaults to `keyof Row` is still not strict enough. Typescript starts spotting errors unless this new generic type is also defined as `extends keyof Row`, which means you can still define only keys of Row, defeating the purpose.

So if you are ok with it, I think I'm going to leave it as is for now. If we find a case in which columns do not strictly match Row fields, there are pieces of logic that would't work anyway, so we can come back to it when we have a concrete case.

Reply:
OK, that makes sense. Let's continue with this as-is.